walkthroughs/week-4-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,104 @@
 
 package com.google.sps;
 
+import java.util.ArrayList;
+import java.util.Arrays;
 import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+import java.util.Set;
+import java.util.HashSet;
 
 public final class FindMeetingQuery {
-  public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
-  }
-}
+    public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
+        if (request.getDuration() > TimeRange.WHOLE_DAY.duration()) {
+            return Arrays.asList();
+        }
+        if (events.isEmpty()) {
+            return Arrays.asList(TimeRange.WHOLE_DAY);
+        }
+
+        int differentAttendees = 1;
+
+        List<TimeRange> available = Arrays.asList(TimeRange.fromStartEnd(TimeRange.START_OF_DAY, TimeRange.END_OF_DAY, true));
+        // Arrays.asList(TimeRange.fromStartEnd(TimeRange.START_OF_DAY, TIME_0800AM, false),
+        //     TimeRange.fromStartEnd(TIME_0830AM, TIME_0900AM, false),
+        //     TimeRange.fromStartEnd(TIME_0930AM, TimeRange.END_OF_DAY, true));
+        for (Event event : events) {
+            if (attendeesOverlap(event.getAttendees(), request.getAttendees())) {
+                available = remove(available, event.getWhen());
+                differentAttendees = 0;
+            }
+        }
+
+        if (differentAttendees == 1) {
+            return Arrays.asList(TimeRange.WHOLE_DAY);
+        }
+        return available;
+    }
+
+    public List<TimeRange> remove(Collection<TimeRange> collectionRanges, TimeRange rangeToRemove) {
```

Comment:
Nice organization of the cases and thanks for including the diagrams. This really helps communicate what each section is handling.

This also reminds me of an old project I used to work on, where we had to do very similar math to adjust prices for a certain time period (e.g. lower the price during the summer then go back to the current price.) We had very very similar diagrams :)